### PR TITLE
feat(optim): use `has_complex` shortcut flag for all applicable optimizers, use `_view_as_real` auxiliary function

### PR DESCRIPTION
**Pull request by jon-chuang**
Follow up to: https://github.com/pytorch/pytorch/pull/110607

CC: @lezcano @janeyx99

### CODE REVIEW

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -125,6 +125,12 @@ def _default_to_fused_or_foreach(params: List[torch.Tensor],
     )
     return fused, foreach
 
+def _view_as_real(params, *state_and_grads):
```
Comment (jon-chuang) [inline torch/optim/optimizer.py:128]: @lezcano looks like we had the same thing in mind

**Review by lezcano — APPROVED** [review]
Nice! This looks much cleaner, thank you.

In the future, you may want to consider using https://github.com/ezyang/ghstack to stack related PRs

**Comment by jon-chuang** [comment]
> In the future, you may want to consider using https://github.com/ezyang/ghstack to stack related PRs

yep just gotta get used to it :stuck_out_tongue:

**Comment by lezcano** [comment]
the failures are real

**Comment by jon-chuang** [comment]
Yes, I'm diagnosing them now (they are a dynamo bug)

**Comment by jon-chuang** [comment]
I'm going to revive this and merge, except for adagrad, which depends on https://github.com/pytorch/pytorch/pull/110709

Will ship an adagrad one after we merge that.

**Pull request closed without merging**